src/login/logind-session.c

```diff
@@ -1226,6 +1229,34 @@ static void session_remove_fifo(Session *s) {
         }
 }
 
+static int session_dispatch_leader_pidfd(sd_event_source *es, int fd, uint32_t revents, void *userdata) {
+        Session *s = ASSERT_PTR(userdata);
+
+        assert(s->leader.fd == fd);
+        session_stop(s, /* force= */ false);
+
+        return 1;
+}
+
+int session_watch_pidfd(Session *s) {
+        int r;
+
+        assert(s);
+        assert(s->leader.fd >= 0);
```

Comment:
as mentioned earlier, i think it would be good if this function would just do an early clean exit if we have no pidfd. i.e. change this assert() here to an `if (!…) return 0`